### PR TITLE
Clear prepared statement cache on idle connection resiliency reconnections

### LINKED ISSUE #1971: SQLServerPreparedStatement uses a wrong handle after server restart - may cause wrong/dangerous statement execution
## Driver version
10.2.1

## SQL Server version
Microsoft SQL Server 2019 (RTM-CU18) (KB5017593) - 15.0.4261.1 (X64)   Sep 12 2022 15:07:06   

## Client Operating System
Windows 10

## JAVA/JVM version
11.0.10

## Table schema
none

## Problem description

SQLServerPreparedStatement.retryBasedOnFailedReuseOfCachedHandle is not properly detecting if the handle is still valid. 
After DB server restart server handle IDs are reset and the numbering starts from 1. So the handle ID from cache may refer to a completely different statement after server restart. The result is unpredictable and may lead to data loss/damaging.

----------sample code-----------------------
public class ZopaSqlTest {

    public static void main(String[] args) throws SQLException, IOException {
        String statement1 = "select SYSTEM_USER where SYSTEM_USER not in (?)";
        String statement2 = "select SYSTEM_USER where SYSTEM_USER not in (?, ?)";
        try(Connection connection = DriverManager.getConnection("jdbc:sqlserver://localhost:1433;databaseName=master;sendStringParametersAsUnicode=false;disableStatementPooling=false;statementPoolingCacheSize=999999;encrypt=false", "username", "password")){

            runPrepared2(connection, statement1, "1");
            runPrepared2(connection, statement2, "1", "2");
            runPrepared2(connection, statement1, "1");
            runPrepared2(connection, statement2, "1", "2");
            System.out.println("Restart DB Server and press enter...");
            System.in.read();
            runPrepared2(connection, statement2, "1", "2");
            runPrepared2(connection, statement1, "1");//here the statement2 is executed instead of statement1
        }
    }

    private static void runPrepared2(Connection connection, String sql, String... params) throws SQLException {
        try(PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            for (int i = 0; i < params.length; i++) {
                preparedStatement.setString(i + 1, params[i]);
            }
            preparedStatement.execute();
        }
    }

}
-----------------------------------

## Expected behavior

After restart the server handle ID should not be used. 

## Actual behavior

The the server handle ID should is used, a wrong statement is executed and a "missing parameter" exception is thrown. 

## Error message/stack trace

Exception in thread "main" com.microsoft.sqlserver.jdbc.SQLServerException: The parameterized query '(@P0 varchar(8000),@P1 varchar(8000))select SYSTEM_USER where SY' expects the parameter '@P1', which was not supplied.
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:265)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1676)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:615)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:537)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7730)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3786)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:268)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:242)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.execute(SQLServerPreparedStatement.java:515)
	at com.edifecs.archival.etl.ZopaSqlTest.runPrepared2(ZopaSqlTest.java:32)
	at com.edifecs.archival.etl.ZopaSqlTest.main(ZopaSqlTest.java:23)

### PR DESCRIPTION
**Pull request by lilgreenbird**
Fixes #1971

### CODE REVIEW

File: src/test/java/com/microsoft/sqlserver/jdbc/resiliency/BasicConnectionTest.java [hunk 1]
```diff
@@ -268,6 +271,47 @@ public void testPooledConnectionLang() throws SQLException {
...
+    /**
+     * Test PreparedStatement Cache is cleared after reconnect
+     * 
+     * @throws SQLException
+     */
+    @Test
+    public void testPreparedStatementCacheShouldBeCleared() throws SQLException {
+        try (SQLServerConnection con = (SQLServerConnection) ResiliencyUtils.getConnection(connectionString)) {
+            int cacheSize = 2;
+            String query = String.format("/*testPreparedStatementCacheShouldBeCleared_%s*/SELECT 1; -- ",
```
Comment (David-Engel) [inline src/test/java/com/microsoft/sqlserver/jdbc/resiliency/BasicConnectionTest.java:283]: I would test a statement that produces multiple result sets:

`select 1;select 1`

Comment (David-Engel) [reply src/test/java/com/microsoft/sqlserver/jdbc/resiliency/BasicConnectionTest.java:283]: Scanning through the list of recent fixes reminded me that we should also cover exception cases. ICR should still work afterwards. Examples:

`select 1/0`
`select 1;select 1/0` <- Processing all result sets and only processing the first set.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 2]
```diff
@@ -3756,6 +3756,10 @@ boolean executeCommand(TDSCommand newCommand) throws SQLServerException {
                                         false);
                             }
                             try {
+                                // Invalidate statement caches in case of server reset
+                                if (null != preparedStatementHandleCache)
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:3760]: Nitpick, brackets for if statement.

File: src/test/java/com/microsoft/sqlserver/jdbc/resiliency/BasicConnectionTest.java [hunk 3]
```diff
@@ -268,6 +271,47 @@ public void testPooledConnectionLang() throws SQLException {
...
+                    UUID.randomUUID().toString());
```
Comment (tkyc) [inline src/test/java/com/microsoft/sqlserver/jdbc/resiliency/BasicConnectionTest.java:284]: toString() isn't necessary here, it can be removed.

**Review by Jeffery-Wasty — APPROVED** [review]

**Review by Jeffery-Wasty — APPROVED** [review]

**Review by tkyc — APPROVED** [review]

**Pull request merged**